MSSQLServer compatibility enhancements

sjednac opened the pull request:
Provides several enhancements for `MSSQLServer` compatibility mode, as described in #1338 (discarding "table hints" & treating "identity" as an `auto_increment` alias). 

The code base changed significantly, since my original "hack", therefore the proposed implementation is brand new, although, the [original idea](http://mintbeans.com/h2-for-sql-server-testing/) didn't change at all.

sjednac committed:
Discard table hints in MSSQLServer mode

Ignore table hints (e.g. "SELECT * FROM table WITH (NOLOCK)"), that form a
valid expression in SQL Server.

Reference:
https://docs.microsoft.com/en-us/sql/t-sql/queries/hints-transact-sql-table?view=sql-server-2017

sjednac committed:
Treat identity as an auto_increment alias in MSSQLServer mode

Add support for `create table test(id int primary key identity)` statements,
which are common in the SQLServer world.

sjednac committed:
Provide a sample showcasing SQL Server compatibility

Add a dedicated example for `MSSQLServer` compatibility mode, that
shows all DBMS-specific features in a central, user-friendly place.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -1855,6 +1860,34 @@ private String readFromAlias(String alias) {
     }
 
+    private void discardWithTableHints() {
+        if (readIf("WITH")) {
```
Use `readIf(WITH)` instead, `WITH` is a keyword with own token type.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -7458,6 +7491,14 @@ private void parseTableColumnDefinition(CommandWithColumns command, Schema schem
                 }
+                if (database.getMode().useIdentityAsAutoIncrement) {
+                    if (readIf("NOT")) {
+                        read("NULL");
```
1. `readIf(NOT)`, `read(NULL)`

2. I guess `column.setNullable(false);` should be invoked here.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -1855,6 +1860,34 @@ private String readFromAlias(String alias) {
+                do {
+                    readExpression();
+                } while (readIf(COMMA));
+                read(CLOSE_PAREN);
```
Use `} while (readIfMore(true));`.

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -1855,6 +1860,34 @@ private String readFromAlias(String alias) {
+            } else if (readIf(EQUAL)){
+                readExpression();
+            } else {
+                throw getSyntaxError();
```
Just use
```Java
} else {
    read(EQUAL);
    readExpression();
}
```

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -1855,6 +1860,34 @@ private String readFromAlias(String alias) {
+    private void discardWithTableHints() {
+        if (readIf("WITH")) {
+            read(OPEN_PAREN);
+            if (!readIf(CLOSE_PAREN)) {
```
`while (readIfMore(true)) {`

katzyn commented on h2/src/main/org/h2/command/Parser.java:
```diff
@@ -1855,6 +1860,34 @@ private String readFromAlias(String alias) {
+        if (readIf("INDEX")) {
+            if (readIf(OPEN_PAREN)) {
+                do {
+                    readExpression();
```
Is `index_value` really an expression? What exactly is allowed here in SQL Server? I cannot find the answer in documentation of SQL Server.

sjednac replied:
I honestly don't know - any suggestions for a `readExpression` replacement?

katzyn replied:
If `readExpression()` works fine with real-world use cases just leave it as is for now.

katzyn commented on h2/src/test/org/h2/samples/SQLServerSample.java:
```diff
@@ -0,0 +1,48 @@
+import java.sql.ResultSet;
+import java.sql.Statement;
+
+public class SQLServerSample {
```
What is the purpose of this sample? It looks like something useless.

If you want to preserve it, please, add a copyright header (copy it from some other source file).

sjednac replied:
You're right - with unit tests in place this is meaningless. I've removed it.

katzyn commented on h2/src/test/org/h2/samples/SQLServerSample.java:
```diff
@@ -0,0 +1,48 @@
+        Statement stat = conn.createStatement();
+
+        stat.execute("create table parent(id int primary key identity, name varchar(255))");
+        stat.execute("create table child(id int primary key identity, name varchar(255), parent_id int, foreign key (parent_id) references public.parent(id) )");
```
This line is too long, only 120 characters for code and 80 for comments are allowed in H2.

katzyn commented on h2/src/test/org/h2/test/db/TestCompatibilitySQLServer.java:
```diff
@@ -0,0 +1,76 @@
+        Statement stat = conn.createStatement();
+
+        stat.execute("create table parent(id int primary key, name varchar(255))");
+        stat.execute("create table child(id int primary key, parent_id int, name varchar(255), foreign key (parent_id) references public.parent(id))");
```
Long line.

katzyn commented on h2/src/test/org/h2/test/db/TestCompatibilitySQLServer.java:
```diff
@@ -0,0 +1,76 @@
+        assertSupportedSyntax(stat, "select * from parent p join child ch on ch.parent_id = p.id");
+        assertSupportedSyntax(stat, "select * from parent p with(nolock) join child ch with(nolock) on ch.parent_id = p.id");
+        assertSupportedSyntax(stat, "select * from parent p with(nolock) join child ch with(nolock, index = id) on ch.parent_id = p.id");
+        assertSupportedSyntax(stat, "select * from parent p with(nolock) join child ch with(nolock, index(id, name)) on ch.parent_id = p.id");
```
Long lines.

katzyn commented on h2/src/test/org/h2/test/db/TestCompatibilitySQLServer.java:
```diff
@@ -0,0 +1,76 @@
+    private void testDiscardTableHints() throws SQLException {
+        deleteDb("sqlserver");
+
+        Connection conn = getConnection("sqlserver;MODE=MSSQLServer");
```
Connection should be closed somewhere. The same problem in `testUseIdentityAsAutoIncrementAlias()`.

Also it's better to delete database not only before execution of test case, but also after it. Actually you can do it in `test()` method and use shared connection in both test cases, there is no reason to use a separate connection in them now.

sjednac replied:
FYI, I actually used `TestCompatibilityOracle` as a reference point (and still forgot to close the connections) - it deletes the db before each test only. Might be worth checking out.

katzyn commented on h2/src/test/org/h2/test/db/TestCompatibilitySQLServer.java:
```diff
@@ -0,0 +1,76 @@
+        assertSupportedSyntax(stat, "create table test2 (id int primary key not null identity)");
+    }
+
+    private void assertSupportedSyntax(Statement stat, String sql) {
```
This method is not needed, just run the SQL directly. If it fails, exception will be visible in log. Your method only hides the stack trace, that's not a good idea.

katzyn commented on h2/src/test/org/h2/test/db/TestCompatibilitySQLServer.java:
```diff
@@ -0,0 +1,76 @@
+package org.h2.test.db;
```
Add a copyright header.

sjednac committed:
Apply code review remarks

Adjust style and remove the `SQLServerSample` as it seems redundant with the
unit test present.

sjednac commented:
Thanks for the comments @katzyn! I've applied some changes and replied to some questions above - please take another look.

katzyn commented:
Please, provide a license statement as described here:
http://h2database.com/html/build.html#providing_patches

It's better to post it to the mailing list:
https://groups.google.com/forum/#!forum/h2-database
Mailing list is partially pre-moderated, new messages may not appear immediately.

sjednac committed:
Add a changelog entry for #1338

sjednac committed:
Add new terms to the build dictionary

Extend dictionary with `discard`, `enhancements` and `nolock` terms.

sjednac committed:
Document new MSSQLServer compatibility features

sjednac commented:
I've posted a license statement to the [mailing list](https://groups.google.com/forum/#!topic/h2-database/jpzepKuM3bI), included a changelog entry, fixed the `spellcheck` target and extended documentation as described [here](http://h2database.com/html/build.html#providing_patches). Thanks for all the tips @katzyn.

katzyn commented:
Thank you for your contribution!

jrsall92 commented:
Can we get a release with this fix if there isn't any?
